elementpath/xpath_nodes.py

```diff
@@ -136,480 +161,755 @@ class NamespaceNode(XPathNode):
-        return hash((self.attribute, self.value))
+    @property
+    def namespaces(self) -> List['NamespaceNode']:
+        if self._namespaces is None:
+            # Lazy generation of namespace nodes of the element
+            position = self.position + 1
+            self._namespaces = [NamespaceNode('xml', XML_NAMESPACE, self, position)]
+            position += 1
+            if self.nsmap:
+                for pfx, uri in self.nsmap.items():
+                    if pfx != 'xml':
+                        self._namespaces.append(NamespaceNode(pfx, uri, self, position))
+                        position += 1
+
+        return self._namespaces
 
+    @property
+    def path(self) -> str:
+        """Returns an absolute path for the node."""
+        path = []
+        item: Any = self
+        while True:
+            if isinstance(item, ElementNode):
+                path.append(item.elem.tag)
+
+            item = item.parent
+            if item is None:
+                return '/{}'.format('/'.join(reversed(path)))
+
+    def match(self, name: str, use_default_namespace: bool = True) -> bool:
+        if name == '*' or name == '*:*':
+            return True
+        elif not name:
+            return not self.name
+        elif name[0] == '*':
+            try:
+                _, _name = name.split(':')
+            except (ValueError, IndexError):
+                raise ElementPathValueError("unexpected format %r for argument 'name'" % name)
+            else:
+                if self.name.startswith('{'):
+                    return self.name.split('}')[1] == _name
+                else:
+                    return self.name == _name
+
+        elif name[-1] == '*':
+            if name[0] != '{' or '}' not in name:
+                raise ElementPathValueError("unexpected format %r for argument 'name'" % name)
+            elif self.name.startswith('{'):
+                return self.name.split('}')[0][1:] == name.split('}')[0][1:]
+            else:
+                return False
+        elif name[0] == '{' or not use_default_namespace:
+            return self.name == name
+        else:
+            default_namespace: Optional[str]
 
-XPathNodeType = Union[ElementNode, DocumentNode, XPathNode]
+            if None in self.nsmap:
+                default_namespace = self.nsmap[None]
+            else:
+                default_namespace = self.nsmap.get('')
+            if not default_namespace:
+                return self.name == name
+            return self.name == '{%s}%s' % (default_namespace, name)
+
+    def iter(self) -> Iterator[XPathNode]:
+        yield self
+
+        iterators: List[Any] = []
+        children: Iterator[Any] = iter(self.children)
+
+        if self._namespaces:
+            yield from self._namespaces
+        if self.attributes:
+            yield from self.attributes
+
+        while True:
+            try:
+                child = next(children)
+            except StopIteration:
```

Comment:
I have refactored the iteration methods with a for loop. Run faster now. 
For comparison with recursive algorithm (if you intended that with nested for-loop) i added an `iter2()` alternative.

Check the performances of both with `python profiling/profile_xpath_nodes.py`

Reply:
If performance is such an important goal, consider compiling some critical modules with Cython in CPython, as lxml does, e.g. with its `lxml._elementpath` module. You'd end up with binary modules in that case (and thus would start distributing binary wheels for some CPython versions), but since they just contain a compiled copy of the Python modules, the binary modules stay completely optional. They'd just make your package run faster in CPython.

Reply:
Performance could be important particularly as XPath processor for _xmlschema_ (identity constraints, assertions), but more important is maintainability.

In the past i tried a cythonization from pure-python with type hints but with no great success (big binary modules, no great improvements), so i suspect that for obtaining great improvements i need to maintain extra code. 
I'm right or wrong on this? Which recipe is used for cythonize `lxml._elementpath.py`?

P.S. in `lxml._elementath.py` there is the comment that addressed me on building this ... :)

Thank you

Reply:
It's probably quite simple, you just need to call `cythonize()` with the list of files that you want to compile:
https://docs.cython.org/en/latest/src/userguide/source_files_and_compilation.html#basic-setup-py
but only on CPython, e.g. this is what lxml does for its .py files
https://github.com/lxml/lxml/blob/dcab10594a2a3bec2f8302f68205dd0204c21c65/setupinfo.py#L19-L31

Whether that makes it (substantially) faster depends on your code. I think I remember that we discussed this back at the time, right?

The ElementPath implementation in ElementTree and lxml is quite fast in Python because it prepares as much as possible ahead of time and then uses a very efficient chain of tiny generators to select nodes from the tree. Compiling that with Cython gives another notable boost, because it drops those generator implementations down into C.

Your critical module seems to be `xpath_token.py`. I'd try compiling only that, and then profile the compiled version and tune it into something that runs fast when compiled (allowing for performance regressions in interpreted CPython). Users who care about performance can then use a binary wheel with CPython, while PyPy will probably still make something fast out of the tuned Python code. Typical changes could be: reorder type checks to test for builtin types before more complex user types, avoid method/function calls and do a bit of inlining instead, maybe convert `XPathToken` (why is that a `MutableSequence`?) and (some of) its subclasses to extension types (`@cython.cclass`) and use type annotations to avoid Python method calls. Not sure if the latter is doable in the current design, though.

Reply:
`XPathToken` is a `MutableSequence` to handle token trees, that is the base of the variation of the TDOP parser that i've implemented for this. Also if i remove MutableSequence as base i've to maintain a set of minimal special methods to let the tokens works as lists. Reworking is a lot of work ...

Other parts for trying now a binary optimization may be the `XPathContext` and the `XPathNode` class hierarchy (dynamic evaluation phase).

XPath has hundreds of symbols to manage (about 200 functions in XPath 3.0) and, apart small optimizations, i prefer to work on implementing the version 3.1 of XPath, that has interfaces with JSON. The new release with a full XPath nodes implementation is preparatory for that.